rfc/0001-shapely-2.0-roadmap.md

```diff
@@ -0,0 +1,586 @@
+  GEOS geometry objects in a Python extension type.
+* Provide functionality to work with arrays of geometry objects with a familiar
+  NumPy array interface, i.e. optimized, vectorized ufuncs for all GEOS
+  operations. This adds a required dependency on numpy.
+* Use the vectorized operations for the implementation of the scalar operations,
+  preserving the existing Shapely scalar API.
+* Make some API changes that are partly required and also desirable (e.g.
+  regarding mutability of geometries, ctypes interface, etc).
+
+In practice, the above proposal means to fully migrate the PyGEOS code into
+Shapely's core, and refactoring the rest of Shapely on top of that (instead
+of taking PyGEOS as a dependency of Shapely).
+
+Note, although this expands the scope of Shapely to *array* functionality,
+this still explicitly excludes feature attribute information (which is left
+to other libraries such as GeoPandas).
+
+On a practical note, the following workflow is proposed:
+
+* The upcoming Shapely 1.8 (the current master branch) includes deprecation
+  warnings for behaviour that will change (in addition to dropping python 2
+  support, [#743](https://github.com/Toblerity/Shapely/issues/743)).
+* Shapely 2.0 becomes the release with the refactor from integrating PyGEOS. A
+  development branch is created for this, which means that initially, we will
+  need to keep the two (or three including 1.7-maint) branches in sync to a
+  certain extent.
+
+
+## Details
+
+### Geometry extension type and Python subclasses
+
+The approach in PyGEOS is to store the pointer to the actual GEOS object in a
+lightweight [Python extension type](https://docs.python.org/3/extending/newtypes_tutorial.html).
+This way, the GEOS pointer is accessible from C without Python overhead as a
+static attribute of the Python object (an attribute of the C struct that makes
+up a Python object).
+
+Hence, a single `Geometry` Python extension type is defined in C wrapping a
+`GEOSGeometry` pointer. In addition, this extension type can further be
+subclassed in Python to provide the geometry type-specific classes (Point,
+LineString, Polygon, etc).
+
+In [pygeos/pygeos#104](https://github.com/pygeos/pygeos/pull/104), we
+experimented with this Python subclassing approach (a registry of subclasses is
+constructed such that the C functions creating new geometries can return the
+correct Python subclasses from C). The conclusion here is that this approach
+only adds minimal overhead to the C functions. This
+[branch](https://github.com/Toblerity/Shapely/compare/master...jorisvandenbossche:geometry-subclasses?expand=1)
+starts to refactor the existing Shapely geometry classes as subclasses of the
+C extension type.
+
+Compared to the current `ctypes` based approach, this has some benefits:
+
+* `ctypes` is not (or no longer) the recommended way to wrap a complex C/C++
+  library. It adds overhead and possible runtime issues.
```

Comment:
Comment from the peanut gallery: do you have some references on this? I assume that the current consensus is to either do a Python extension type or to use CFFI, but it would be great to have some reading material on the topic.

Reply:
@astrojuanlu sorry for the *very*late reply. I don't have a direct reference at hand that explains this (more based on general experience and from talking to other people), but we certainly had several reports with installation issues related to this.